dll/win32/shimgvw/shimgvw.c

```diff
@@ -41,6 +39,108 @@ WNDPROC PrevProc = NULL;
 
 HWND hDispWnd, hToolBar;
 
+/* zooming */
+#define MIN_ZOOM 10
+#define MAX_ZOOM 1600
+UINT ZoomPercents = 100;
+static const UINT ZoomSteps[] =
+{
+    10, 25, 50, 100, 200, 400, 800, 1600
+};
+
+static void ZoomIn(void)
+{
+    INT i;
+
+    /* find next step */
+    for (i = 0; i < ARRAYSIZE(ZoomSteps); ++i)
+    {
+        if (ZoomPercents < ZoomSteps[i])
+            break;
+    }
+    if (i == ARRAYSIZE(ZoomSteps))
+        ZoomPercents = MAX_ZOOM;
+    else
+        ZoomPercents = ZoomSteps[i];
+
+    /* update tool bar buttons */
+    SendMessage(hToolBar, TB_ENABLEBUTTON, IDC_ZOOMM, TRUE);
+    if (ZoomPercents >= MAX_ZOOM)
+        SendMessage(hToolBar, TB_ENABLEBUTTON, IDC_ZOOMP, FALSE);
+    else
+        SendMessage(hToolBar, TB_ENABLEBUTTON, IDC_ZOOMP, TRUE);
+
+    /* redraw */
+    InvalidateRect(hDispWnd, NULL, TRUE);
+}
+
+static void ZoomOut(void)
```

Comment:
These two functions can be merged into a single one, with a parameter indicating which "direction" (zooming in or out) to do

Reply:
Done.